### ISSUE TITLE
BlockOwnerDeletion does not block owner deletion

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#86509]
I created a CR "parent" that creates a CR "child".  The "child" CR has **two** ownerrefs: "parent" and "uncle" (another CR).  The "parent" ownerRef blockOwnerDeletion is true; the "uncle" ownerRef blockOwnerDeletion is unset.

When I delete "parent", "child" is not deleted.  We do not seem to honor blockOwnerDeletion on "child" being deleted, though we do remove the ownerRef.  If I first remove the "uncle" ownerRef,  "child"  is deleted when I delete "parent".

Is this expected behaviour?  It seems contrary to all docs I have found (and the field name!)

I would expect: that either child would be deleted, or that the parent deletion would hang (if the second ownerRef was considered blocking) until "child" has been deleted.

**Comment by liggitt** [comment]
> When I delete "parent", "child" is not deleted. 

That is expected. A child is not deleted by GC until all parents are absent. 

> We do not seem to honor blockOwnerDeletion on "child" being deleted, though we do remove the ownerRef.

That is surprising to me. Are you deleting the parent with an "orphan dependents" deletion option? If not, I would expect the ownerRef to remain and for blockOwnerDeletion to be effective.

**Comment by caesarxuchao** [comment]
> When I delete "parent", "child" is not deleted. We do not seem to honor blockOwnerDeletion on "child" being deleted, though we do remove the ownerRef.

Are you saying that when you delete the parent, the "parent" ownerRef in the child object got removed?

**Comment by sethp-nr** [comment]
I was able to reproduce this issue with the core Secret type as well:

```
kubetl create secret generic parent
kubetl create secret generic child
kubetl create secret generic uncle
```

and then arrange the child's owner references in the way @justinsb described:

```
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: "2020-03-03T00:12:29Z"
  name: child
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    kind: Secret
    name: parent
    uid: 1bb00bb8-99d0-4dee-9319-08f75fdcb93b
  - apiVersion: v1
    kind: Secret
    name: uncle
    uid: 81f7727a-6ed6-4c35-aca6-3411016ad3d0
  resourceVersion: "50458"
  selfLink: /api/v1/namespaces/sp1a/secrets/child
  uid: 01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146
type: Opaque
```

and then `kubectl delete secret parent` does _not_ delete `child` but the "blocking" owner reference is removed:

```
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: "2020-03-03T00:12:29Z"
  name: child
  namespace: sp1a
  ownerReferences:
  - apiVersion: v1
    kind: Secret
    name: uncle
    uid: 81f7727a-6ed6-4c35-aca6-3411016ad3d0
  resourceVersion: "49534"
  selfLink: /api/v1/namespaces/sp1a/secrets/child
  uid: 01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146
type: Opaque
```


via `kubectl proxy &` and `curl localhost:8001/api/v1/namespaces/default/secrets?watch=1` we can get a "play by play":

```
{"type":"ADDED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"parent","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/parent","uid":"1bb00bb8-99d0-4dee-9319-08f75fdcb93b","resourceVersion":"48141","creationTimestamp":"2020-03-03T00:12:27Z"},"type":"Opaque"}}
{"type":"ADDED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"child","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/child","uid":"01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146","resourceVersion":"48147","creationTimestamp":"2020-03-03T00:12:29Z"},"type":"Opaque"}}
{"type":"ADDED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"uncle","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/uncle","uid":"81f7727a-6ed6-4c35-aca6-3411016ad3d0","resourceVersion":"48170","creationTimestamp":"2020-03-03T00:12:32Z"},"type":"Opaque"}}
{"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"child","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/child","uid":"01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146","resourceVersion":"49465","creationTimestamp":"2020-03-03T00:12:29Z","ownerReferences":[{"apiVersion":"v1","kind":"Secret","name":"parent","uid":"1bb00bb8-99d0-4dee-9319-08f75fdcb93b","blockOwnerDeletion":true},{"apiVersion":"v1","kind":"Secret","name":"uncle","uid":"81f7727a-6ed6-4c35-aca6-3411016ad3d0"}]},"type":"Opaque"}}
{"type":"DELETED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"parent","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/parent","uid":"1bb00bb8-99d0-4dee-9319-08f75fdcb93b","resourceVersion":"49529","creationTimestamp":"2020-03-03T00:12:27Z"},"type":"Opaque"}}
{"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"child","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/child","uid":"01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146","resourceVersion":"49534","creationTimestamp":"2020-03-03T00:12:29Z","ownerReferences":[{"apiVersion":"v1","kind":"Secret","name":"uncle","uid":"81f7727a-6ed6-4c35-aca6-3411016ad3d0"}]},"type":"Opaque"}}
```

Note that the `parent` is DELETED without any modification (as would be expected for an `orphan` finalizer add/remove), and the `child` is MODIFIED but never DELETED.

Further, using the `curl` from the [garbage collection docs](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#setting-the-cascading-deletion-policy) to set the (non-default) foreground deletion policy that's [required for owner's deletion to wait for its children flagged with `blockOwnerDeletion` to disappear](https://gist.github.com/sethp-nr/bd9495bde5e136d2f0fb78609a527486) produces an almost identical result (initial adds and modifies omitted):

```
{"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"parent","namespace":"sp1a","selfLink":"/api/v1/namespaces/sp1a/secrets/parent","uid":"f0d3d9bb-4345-445b-8830-4210fa729389","resourceVersion":"51848","creationTimestamp":"2020-03-03T00:19:19Z","deletionTimestamp":"2020-03-03T00:24:27Z","deletionGracePeriodSeconds":0,"finalizers":["foregroundDeletion"]},"type":"Opaque"}}
{"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"child","namespace":"sp1a","selfLink":"/api/v1/namespaces/sp1a/secrets/child","uid":"01d74e61-a4ce-4cb3-98a9-bbbc8d9ab146","resourceVersion":"51851","creationTimestamp":"2020-03-03T00:12:29Z","ownerReferences":[{"apiVersion":"v1","kind":"Secret","name":"uncle","uid":"81f7727a-6ed6-4c35-aca6-3411016ad3d0"}]},"type":"Opaque"}}
{"type":"DELETED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"parent","namespace":"sp1a","selfLink":"/api/v1/namespaces/sp1a/secrets/parent","uid":"f0d3d9bb-4345-445b-8830-4210fa729389","resourceVersion":"51852","creationTimestamp":"2020-03-03T00:19:19Z","deletionTimestamp":"2020-03-03T00:24:27Z","deletionGracePeriodSeconds":0,"finalizers":["foregroundDeletion"]},"type":"Opaque"}}
```

Now the order is right (the child is MODIFIED before the parent is DELETED) but the behavior is still to remove the owner ref instead of deleting the child.

**Comment by tedyu** [comment]
I am on mobile device. 
From what I read from above, the presence of secret ‘uncle’ was the cause for child not being deleted.

**Comment by sethp-nr** [comment]
Oh, I see – I was thinking of the `blockOwnerDeletion` owner references as an "OR" (if either owner is deleted, so is the dependent), but it sounds like apimachinery is thinking of them as an "AND" operation, at least in `background` deletion mode.

Is there a way to mark a resource as "foreground deletion by default", or some other way to flag a delete as something that should cascade (& block)? It seems unfortunate that the cluster api controllers have to re-implement a lot of the garbage collector's functionality in order to get back to "OR" semantics.

**Comment by avorima** [comment]
>> I would expect: that either child would be deleted, or that the parent deletion would hang (if the second ownerRef was considered blocking) until "child" has been deleted.

>The second behavior is what I would expect.

I have a question reguarding this. Here's my setup:

```
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: "2020-03-19T13:34:11Z"
  name: parent
  namespace: default
  resourceVersion: "1652"
  selfLink: /api/v1/namespaces/default/secrets/parent
  uid: 0b4845a3-b6bb-4a1c-b024-3e1d65593e39
type: Opaque
```

 ```
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: "2020-03-19T13:34:18Z"
  finalizers:
  - dont.delete/me
  name: child
  namespace: default
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    kind: Secret
    name: parent
    uid: 0b4845a3-b6bb-4a1c-b024-3e1d65593e39
  resourceVersion: "1819"
  selfLink: /api/v1/namespaces/default/secrets/child
  uid: 7384cd73-97a6-47a5-826b-05f46a4eecab
type: Opaque
```

```
kubectl delete secret parent
```

I get the following events:
```
{"type":"DELETED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"parent","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/parent","uid":"0b4845a3-b6bb-4a1c-b024-3e1d65593e39","resourceVersion":"1955","creationTimestamp":"2020-03-19T13:34:11Z"},"type":"Opaque"}}
{"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1","metadata":{"name":"child","namespace":"default","selfLink":"/api/v1/namespaces/default/secrets/child","uid":"7384cd73-97a6-47a5-826b-05f46a4eecab","resourceVersion":"1956","creationTimestamp":"2020-03-19T13:34:18Z","deletionTimestamp":"2020-03-19T13:35:59Z","deletionGracePeriodSeconds":0,"ownerReferences":[{"apiVersion":"v1","kind":"Secret","name":"parent","uid":"0b4845a3-b6bb-4a1c-b024-3e1d65593e39","blockOwnerDeletion":true}],"finalizers":["dont.delete/me"]},"type":"Opaque"}}
```

So even though the `deletionTimestamp` is set on the child, the owner deletion does not wait until the finalizer is removed.

Is this expected behaviour? And if so, is it possible to ensure that the owner waits until the child is actually deleted?

**Comment by avorima** [comment]
Ok, so the issue is that `kubectl delete` apparently doesn't actually set the deletion propagation to foreground. I would expect this to be the case if `--wait` and `--cascade` are set to `true`.

**Comment by derekperkins** [comment]
We are seeing this exact behavior as well, with two CRDs. It is definitely misleading and I believe wrong that `blockOwnerDeletion` doesn't actually prevent the owner from being deleted.
cc @whizard

**Comment by ymmt2005** [comment]
/remove-lifecycle rotten

Has anyone been working on this?

**Comment by avorima** [comment]
@ymmt2005 This issue was fixed in Kubernetes 1.20. If you rerun my test setup with `--cascade=foreground` the parent secret is not deleted until the child secret's finalizer is removed.

**Comment by ymmt2005** [comment]
@avorima Thank you for sharing it!

**Comment by caesarxuchao** [comment]
The garbage collection behavior described in https://github.com/kubernetes/kubernetes/issues/86509#issue-541346977, https://github.com/kubernetes/kubernetes/issues/86509#issuecomment-601188296 are as expected.

When using kubectl to delete the parent, to make the blockOwnerverDeletion option of the child's ownerRef effective, one needs to set `--cascade=foreground`.

**Comment by caozhuozi** [comment]
a summary for future comers:
1. Regardless of the deletion policy (orphan, foreground, or background), a child resource cannot be deleted if there are any other owner references attached to it. Instead, only the corresponding owner reference is removed.
2. `blockOwnerDeletion` only takes effect when the delete policy is `foreground`. I think a more clear name might be `IfInForgroundModeBlockOwnerDeletion`. Specifically,
    * if you delete a parent resource in `orphan` or `background` mode, the `blockOwnerDeletion` options in child resources will be ignored and the parent will be deleted without being blocked (also assuming there are no finalizers).
    * if you delete a parent resource in `forground` mode, but if no child resource has `blockOwnerDeletion` being `true`, the parent will also be deleted without being blocked.
    * When a parent is being delelted in `foreground` mode, the garbage collector will look for child resources with `blockOwnerDeletion` set to true. If it cannot delete these child resources due to other owner references (as described in point 1), it will remove the owner reference instead. Once all blocked child resources are either deleted or have their `blockOwnerDeletion` owner reference removed, the parent can then be deleted by the GC.

**Comment by Zeel-Patel** [comment]
@caozhuozi, if parent is deleted without foreground mode, will child be eventually deleted by some GC? Consider child only has single parent owner ref.

**Comment by ian-howell** [comment]
Was this supposed to be closed?

From @caozhuozi's response:
> blockOwnerDeletion only takes effect when the delete policy is foreground

I verified this using this script:

```
#!/bin/bash

kubectl create ns secret-owner-ref-demo

kubectl apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: owner
  namespace: secret-owner-ref-demo
type: Opaque
EOF

OWNER_UID=$(kubectl get secret owner -o jsonpath='{.metadata.uid}') && kubectl apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  name: dependent
  namespace: secret-owner-ref-demo
  finalizers:
    - kubernetes.io/secret-protection
  ownerReferences:
    - apiVersion: v1
      kind: Secret
      name: owner
      uid: $OWNER_UID
      blockOwnerDeletion: true
      controller: false
type: Opaque
EOF
```

Here's the results:

| `blockOwnerDeletion` | delete command | result |
|-|-|-|
| false | kubectl delete secrets owner | owner is deleted immediately | 
| true | kubectl delete secrets owner | owner is deleted immediately | 
| false | kubectl delete secrets owner --cascade=foreground  | owner is deleted immediately | 
| true | kubectl delete secrets owner --cascade=foreground  | owner does **not** get deleted | 

So in other words, in order to get foreground deletion, you need _both_ `blockOwnerDeletion` to be true **and** the delete policy has to be `foreground`? And if either of those conditions aren't met, we default to background deletion?

Does that sound right? It feels like it should be an "OR" of the 2 conditions, and not an "AND". Considering that this GH issue is the second google result for "blockownerdeletion" (after the official documentation), I think it's safe to say that I'm not the only one who is confused by this.
